go-controller/pkg/ovn/controller/apbroute/external_controller.go

```diff
@@ -16,154 +18,198 @@ import (
-				continue
+// InsertOverwrite should be used to add elements to the gatewayInfoList.
+// The latest added gateway with duplicate ip will cause existing gw ip to be deleted.
+// This way, we always have only 1 gatewayInfo for every ip.
+func (g *gatewayInfoList) InsertOverwrite(gws ...*gatewayInfo) {
+	for _, gw := range gws {
+		g.insertOverwrite(gw)
+	}
+}
+
+func (g *gatewayInfoList) insertOverwrite(gw *gatewayInfo) {
+	if len(gw.Gateways) == 0 {
+		return
+	}
+	emptyIdxs := []int{}
+	for idx, existingGW := range g.elems {
+		if existingGW.Equal(gw) {
+			// gw already exists in the gatewayInfoList
+			return
+		}
+		// make sure duplicate ips only exist in the latest-added gatewayInfo
+		existingGW.RemoveIPs(gw)
+		if len(existingGW.Gateways) == 0 {
+			// all existingGW ips are overwritten by gw
+			emptyIdxs = append(emptyIdxs, idx)
+		}
+	}
+
+	g.remove(emptyIdxs...)
+	g.elems = append(g.elems, gw)
+}
+
+func (g *gatewayInfoList) InsertOverwriteFailed(gws ...*gatewayInfo) {
+	for _, gw := range gws {
+		gw.failedAdd = true
+	}
+	g.InsertOverwrite(gws...)
+}
+
+// Delete removes gatewayInfos that match for all fields, including failedAdd status
+func (g *gatewayInfoList) Delete(gws ...*gatewayInfo) {
+	for _, gw := range gws {
+		deleteIdx := 0
+		found := false
+		for idx, i := range g.elems {
+			if i.Equal(gw) {
+				found = true
+				deleteIdx = idx
+				break
 			}
-			gws.Insert(ip)
 		}
-		if len(gws) > 0 {
-			ret = append(ret, newGatewayInfo(gws, item.BFDEnabled))
+		if found {
+			g.remove(deleteIdx)
 		}
 	}
-	return ret, duplicated, nil
 }
-func (g gatewayInfoList) Delete(item *gatewayInfo) gatewayInfoList {
-	ret := gatewayInfoList{}
-	for _, i := range g {
-		if !i.Equal(item) {
-			ret, _, _ = ret.Insert(i)
+
+func (g *gatewayInfoList) remove(idxs ...int) {
+	if len(idxs) == 0 {
+		return
+	}
+	newElems := make([]*gatewayInfo, 0, len(g.elems))
+	idxToDelete := sets.New[int](idxs...)
+	for existingIdx, existingElem := range g.elems {
+		if !idxToDelete.Has(existingIdx) {
+			newElems = append(newElems, existingElem)
 		}
 	}
-	return ret
+	g.elems = newElems
 }
 
-func (g gatewayInfoList) Len() int {
-	return len(g)
+func (g *gatewayInfoList) Len() int {
+	return len(g.elems)
+}
+
+// Equal compares gatewayInfoList elements to be exactly the same, including failedAdd status,
+// but ignores elements order.
+func (g *gatewayInfoList) Equal(g2 *gatewayInfoList) bool {
+	if len(g.elems) != len(g2.elems) {
+		return false
+	}
+	for _, e1 := range g.elems {
+		// since gatewayInfoList shouldn't have elements with the same ips, it is safe to assume
+		// every element can be equal to only element of another gatewayInfoList
+		found := false
+		for _, e2 := range g2.elems {
+			if e1.Equal(e2) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
 }
 
 type gatewayInfo struct {
-	Gateways   *syncSet
+	Gateways   sets.Set[string]
 	BFDEnabled bool
+	failedAdd  bool
```

Comment:
I'm unclear what is the benefit of keeping track of which gw IPs were successfully applied and which were not if we don't do any cleanup in case of failure while applying them. 
Since operations with the libovsdb should be idempotent, re-applying or deleting should not impact the functionality. Is it to avoid the 50ms penalty interacting with the DB?

Reply:
yes, this field allows to skip db transaction for already applied changes, usually the db transaction delay is one of the biggest waiting times during event handling, also it will be multiplied by the number of target pod, since every pod will be configured with its own transaction. So I would say it is simple enough to save some transactions